src/Servers/HttpSys/test/FunctionalTests/ResponseTrailersTests.cs

```diff
@@ -0,0 +1,384 @@
+                }
+                return Task.FromResult(0);
+            }))
+            {
+                var response = await SendRequestAsync(address);
+                response.EnsureSuccessStatusCode();
+                Assert.Equal(HttpVersion.Version20, response.Version);
+                Assert.Empty(response.TrailingHeaders);
+            }
+        }
+
+        [ConditionalFact]
+        [MinimumOSVersion(OperatingSystems.Windows, "10.0.19505", SkipReason = "Requires HTTP/2 Trailers support.")]
+        public async Task ResponseTrailers_NoBody_TrailersSent()
+        {
+            using (Utilities.CreateDynamicHttpsServer(out var address, httpContext =>
+            {
+                httpContext.Response.DeclareTrailer("trailername");
+                httpContext.Response.AppendTrailer("trailername", "TrailerValue");
+                return Task.FromResult(0);
+            }))
+            {
+                var response = await SendRequestAsync(address);
+                response.EnsureSuccessStatusCode();
+                Assert.Equal(HttpVersion.Version20, response.Version);
+                Assert.NotEmpty(response.TrailingHeaders);
+                Assert.Equal("TrailerValue", response.TrailingHeaders.GetValues("TrailerName").Single());
+            }
+        }
+
+        [ConditionalFact]
+        [MinimumOSVersion(OperatingSystems.Windows, "10.0.19505", SkipReason = "Requires HTTP/2 Trailers support.")]
+        public async Task ResponseTrailers_WithBody_TrailersSent()
+        {
+            using (Utilities.CreateDynamicHttpsServer(out var address, async httpContext =>
+            {
+                await httpContext.Response.WriteAsync("Hello World");
+                httpContext.Response.AppendTrailer("TrailerName", "Trailer Value");
+            }))
+            {
+                var response = await SendRequestAsync(address);
+                response.EnsureSuccessStatusCode();
+                Assert.Equal(HttpVersion.Version20, response.Version);
+                Assert.Equal("Hello World", await response.Content.ReadAsStringAsync());
+                Assert.NotEmpty(response.TrailingHeaders);
+                Assert.Equal("Trailer Value", response.TrailingHeaders.GetValues("TrailerName").Single());
+            }
+        }
+
+        [ConditionalFact]
+        [MinimumOSVersion(OperatingSystems.Windows, "10.0.19505", SkipReason = "Requires HTTP/2 Trailers support.")]
+        public async Task ResponseTrailers_WithContentLengthBody_TrailersNotSent()
+        {
+            var body = "Hello World";
+            using (Utilities.CreateDynamicHttpsServer(out var address, async httpContext =>
+            {
+                httpContext.Response.ContentLength = body.Length;
+                await httpContext.Response.WriteAsync(body);
+                httpContext.Response.AppendTrailer("TrailerName", "Trailer Value");
```

Comment:
If I understand it correctly, there are no trailers set because AppendTrailer was set after WriteAsync fulfilled the content length requirement? Add a comment?

Reply:
Correct. The server had no expectation any trailers would be added so it ended the response by default.